pkg/controller/drain/drain_controller.go

```diff
@@ -0,0 +1,417 @@
+
+	glog.Info("Starting MachineConfigController-DrainController")
+	defer glog.Info("Shutting down MachineConfigController-DrainController")
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(ctrl.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (ctrl *Controller) handleNodeEvent(node interface{}) {
+	n := node.(*corev1.Node)
+	glog.V(4).Infof("Updating Node %s", n.Name)
+	ctrl.enqueueNode(n)
+}
+
+func (ctrl *Controller) enqueue(node *corev1.Node) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(node)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", node, err))
+		return
+	}
+
+	ctrl.queue.Add(key)
+}
+
+func (ctrl *Controller) enqueueRateLimited(node *corev1.Node) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(node)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", node, err))
+		return
+	}
+
+	ctrl.queue.AddRateLimited(key)
+}
+
+// enqueueAfter will enqueue a pool after the provided amount of time.
+func (ctrl *Controller) enqueueAfter(node *corev1.Node, after time.Duration) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(node)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", node, err))
+		return
+	}
+
+	ctrl.queue.AddAfter(key, after)
+}
+
+// enqueueDefault calls a default enqueue function
+func (ctrl *Controller) enqueueDefault(node *corev1.Node) {
+	ctrl.enqueueAfter(node, updateDelay)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (ctrl *Controller) worker() {
+	for ctrl.processNextWorkItem() {
+	}
+}
+
+func (ctrl *Controller) processNextWorkItem() bool {
+	key, quit := ctrl.queue.Get()
+	if quit {
+		return false
+	}
+	defer ctrl.queue.Done(key)
+
+	err := ctrl.syncHandler(key.(string))
+	ctrl.handleErr(err, key)
+
+	return true
+}
+
+func (ctrl *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		ctrl.queue.Forget(key)
+		return
+	}
+
+	if ctrl.queue.NumRequeues(key) < maxRetries {
+		glog.V(2).Infof("Error syncing node %v: %v", key, err)
+		ctrl.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	glog.V(2).Infof("Dropping node %q out of the queue: %v", key, err)
+	ctrl.queue.Forget(key)
+	ctrl.queue.AddAfter(key, 1*time.Minute)
+}
+
+func (ctrl *Controller) syncNode(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Started syncing node %q (%v)", key, startTime)
+	defer func() {
+		glog.V(4).Infof("Finished syncing node %q (%v)", key, time.Since(startTime))
+	}()
+
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	// First check if a drain request is needed for this node
+	// TODO (jerzhang): consider not using the cache for this,
+	// since it has been shown to be problematic in the past
```

Comment:
Curious about that

Reply:
Our current uncordon logic doesn't actually work very well for 2 reasons:
1. minor bug in our logic that we remove the file before success, so if we timeout the first sync for some reason, we never uncordon
2. the uncordon checks we have don't work since the cache isnt updated fast enough (via nodelister) so we are always reading a stale status (on whether the node is successfully uncordoned or not)